design/baremetal-operator/limit-hosts-provisioning.md

```diff
@@ -0,0 +1,121 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# limit-hosts-provisioning
+
+## Status
+
+provisional
+
+## Summary
+
+This proposal aims to introduce a limit to the number of hosts that the
+BareMetal Operator can provision or register simultaneously as a safety
+measure to avoid overloading excessively the system
+
+## Motivation
+
+In the current implementation BMO takes care of handling the complete
+provisioning cycle for a bare metal host without considering the pressure
+level applied to the underlying Ironic services (DHCP included).
+A user that desires to scale up several hundreds of nodes in a single step
+could suffer issues produced from the sudden resources depletion
+
+### Goals
+
+1. Avoid any system instability due the excessive number of hosts being
+   provisioned simultaneously
+
+### Non-Goals
+
+1. Optimize the deployment time of large batches of nodes
+2. Allow the user to configure such behavior
+3. Modify the current Ironic service layout
+
+## Proposal
+
+BMO reconciliation loop already checks for Ironic availability before
+processing an update and the request gets rescheduled if Ironic is not ready.
+After that, a new check will verify if Ironic is overloaded or not so that
+the request could be rescheduled or processed as usual.
```

Comment:
Does that mean a change in the Provisioner interface? What will that look like?

Reply:
My initial idea was not to change the Provisioner interface, and to embed the implementation of this new check directly within the `IsReady` method: from the point of view of the controller, it could make sense to have the provisioner "not ready" to process the current update (since its workload is too high) and to reschedule the request. In this way, only the `ironic` package will be affected, and the current reconcile loop will remain unchanged.

Reply:
OK, that makes sense.